### PR TITLE
Port QuickJS to SideVM

### PR DESCRIPTION
**Pull request by kvinwang**
# Description
This PR ports QuickJS to SideVM and adds support for async IO.
There are a few  implemented async APIs includes `setTimeout/setInterval`, `fetch`, `XMLHttpRequest`, `httpListen`.
See `examples/` for features.

# The execution lifetime

As a SideVM program, it need to be deployed by a pink contract.
At startup, `sidevm-quickjs` would query to the host pink contract (by `sidevm_init_script`) to get an entry js code to run. At any time when it running, the host pink contract can send more js code snippets to sidevm-quickjs instance to eval.
A sidevm-quickjs instance can host multiple js vm instances, each js instance is associated with it's name. The host pink contract can reset/destroy js instances at any time via sending message.

See the e2e test `tests/tests/itWorks.ts` for detail.

# Related changes in pRuntime

This PR requires some changes to pRuntime to be applied:
https://github.com/Phala-Network/phala-blockchain/pull/1366: for listening incoming HTTP request via pRPC endpoint.
https://github.com/Phala-Network/phala-blockchain/pull/1398: for query init script from pink contract.
https://github.com/Phala-Network/phala-blockchain/pull/1399: for charging sidevm instance and deploying to selected workers in transaction.

### CODE REVIEW

File: tests/devphase.config.ts
Comment (h4x3rotab) [inline tests/devphase.config.ts:1]: Let's upgrade to the latest devphase. It's not hard actually.

File: src/service_keeper.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        match service.exec_script(source) {
+            Ok(_) => {}
+            Err(err) => {
+                error!("Executing script [{name}] returned error: {err}");
+            }
+        }
+    }
+
+    pub fn handle_query(_from: Option<AccountId>, query: &[u8]) -> Vec<u8> {
+        if String::from_utf8_lossy(query) == "ping" {
```
Comment (h4x3rotab) [inline src/service_keeper.rs]:
```suggestion
        if query == b"ping" {
```

**Review by Leechael — APPROVED** [review]
lgtm

File: bootcode/js/src/polyfill-url.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                result += this.search;
+            }
+            if (this.hash) {
+                result += this.hash;
+            }
+            return result;
+        }
+
+    }
+    g.URLSearchParams = class URLSearchParams {
```
Comment (Leechael) [inline bootcode/js/src/polyfill-url.js]: The `URLSearchParams` broken on this case:
```js
const a = new URLSearchParams()
a.append('foo', '1')
a.append('foo', '2')
a.toString() === 'foo=1&foo=2'
```

**Comment by Leechael** [comment]
~~I suggest added a way for writing test cases for polyfill, since there are lots of edge cases not yet covered (e.g. `URL` & `URLSearchParams`).~~

Updated: considering using [core-js](https://github.com/zloirock/core-js) for polyfill since it always implemented almost all standard APIs.

File: examples/httpListen.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+console.log = Sidevm.inspect;
```
Comment (h4x3rotab) [inline examples/httpListen.js:1]: Why do we need this?

File: examples/incomingFetch.js
Comment (h4x3rotab) [inline examples/incomingFetch.js:1]: What does this example show?

File: src/service/resource.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    type Error = ValueError;
+
+    fn try_from(v: js::Value) -> Result<Self, Self::Error> {
+        match &v {
+            js::Value::Undefined => Ok(OwnedJsValue::Undefined),
+            js::Value::Null => Ok(OwnedJsValue::Null),
+            js::Value::Exception => Ok(OwnedJsValue::Exception),
+            js::Value::Other { ctx, value } => {
+                let runtime =
+                    js_context_get_runtime(&ctx).ok_or(ValueError::Static("service gone"))?;
```
Comment (h4x3rotab) [inline src/service/resource.rs]: Let's have a unified ValueError for this?

File: bootcode/js/src/polyfill-url.js (hunk 2 above)
Comment (kvinwang) [reply bootcode/js/src/polyfill-url.js]: Replaced with core-js.

File: tests/bind.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    return ctx.paths.currentStack;
+}
+
+async function main() {
+    const stackDir = await prepareStack();
+    bind(`${stackDir}/system.contract`, 'System');
+    bind(`${stackDir}/sidevm_deployer.contract`, 'SidevmDeployer');
+}
+
+main().catch(console.error);
```
Comment (h4x3rotab) [inline tests/bind.ts]:
```suggestion
main().then(() => process.exit(0)).catch(console.error).finally(() => process.exit(-1));
```

Otherwise the lifecycle of async function may have weird behavior

File: tests/bind.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { RuntimeContext, RunMode, AbiTypeBindingProcessor, TypeBinder } from "@devphase/service";
+import * as fs from 'fs';
+
+function bind(contract: string, name: string) {
```
Comment (h4x3rotab) [inline tests/bind.ts]: `bind` is a builtin name. Better to change to `createBinding`

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    pub use sidevm::main;
+
+    pub fn http_connector() -> HttpConnector {
+        HttpConnector::new()
+    }
+
+    async fn get_init_script() -> Option<String> {
+        type LangError = u8;
+        let myid = sidevm::ocall::vmid().ok()?;
```
Comment (h4x3rotab) [inline src/lib.rs]: May those `.ok()` hide too many error reasons? Probably better to write an inner function returning `Result<String>` and then convert to `Option<String` outside?

File: src/host_functions/http_request.rs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    let headers: BTreeSet<&str> = req.headers.pairs.iter().map(|(k, _v)| k.as_str()).collect();
+    // Append Host, Content-Length and Content-Length if not present
+    if !headers.contains("Host") {
+        builder = builder.header("Host", uri.host().unwrap_or_default());
+    }
+    if !headers.contains("Content-Length") {
+        builder = builder.header("Content-Length", body.len());
+    }
+    if !headers.contains("User-Agent") {
+        builder = builder.header("User-Agent", "sidevm-quickjs/0.1.0");
```
Comment (h4x3rotab) [inline src/host_functions/http_request.rs]: Add phat contract like it in pink?

File: src/host_functions/timer.rs [hunk 9]
```diff
@@ -0,0 +1,55 @@
+use super::*;
+use crate::{runtime::time::sleep, service::OwnedJsValue};
+
+pub(crate) fn setup(ns: &js::Value) -> Result<()> {
```
Comment (h4x3rotab) [inline src/host_functions/timer.rs:4]: Suggested comment:

```
/// `clearTimeout` and `clearInterval` are implemented by `close` on the guest side
```

File: src/service_keeper.rs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                if reset {
+                    Self::reset(&name);
+                }
+                Self::exec_script(&name, &source);
+            }
+            Message::Reset { name } => Self::reset(&name),
+        }
+    }
+
+    pub fn handle_connection(connection: crate::runtime::HttpRequest) -> Result<()> {
```
Comment (h4x3rotab) [inline src/service_keeper.rs:94]: Maybe add an example path to the docstring?

File: src/host_functions/http_listen.rs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        opaque_input_stream: js::Value::new_opaque_object(service.context(), input_stream),
+        opaque_output_stream: js::Value::new_opaque_object(service.context(), output_stream),
+    };
+    if let Err(err) = service.call_function(listener, (req,)) {
+        anyhow::bail!("Failed to fire http request event: {err}");
+    }
+    Ok(())
+}
+
+fn use_2nd<F1, I1, F2, O>(_f1: F1, f2: F2) -> Option<O>
```
Comment (h4x3rotab) [inline src/host_functions/http_listen.rs]: What does this function do, for simplifying the type definition?

File: src/host_functions/http_listen.rs [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+where
+    F1: FnOnce(I1) -> O,
+    F2: FnOnce() -> Option<O>,
+{
+    f2()
+}
+
+#[js::host_call]
+fn http_send_response(tx: js::Value, response: HttpResponseHead) {
```
Comment (h4x3rotab) [inline src/host_functions/http_listen.rs]: Should we call it `http_send_response_head`, since it doesn't take care of the body?

File: examples/httpListen.js (hunk 3 above)
Comment (kvinwang) [reply examples/httpListen.js:1]: The `Sidevm.inspect` recursively inspects the object, for example:
```js
const head = {
    statusCode: 200,
    headers: {
        "Foo": "Bar",
        "Content-Length": "123"
    }
};

console.log("console.log head:", head);
Sidevm.inspect("Sidevm.inspect head:", head);
```
Whould output:
```
console.log head: [object Object]
Sidevm.inspect head: {
  statusCode: 200,
  headers: {
    Foo: "Bar",
    Content-Length: "123"
  }
}
```

File: examples/incomingFetch.js
Comment (kvinwang) [reply examples/incomingFetch.js:1]: It shows handling incoming HTTP Request in CF workers style API. (PoC)

File: src/host_functions/http_listen.rs (hunk 11 above)
Comment (kvinwang) [reply src/host_functions/http_listen.rs]: Yes.
It allow the compiler to infer the type of the second expression from the first one. For instance: 
```rust
    let Some(read_half) = use_2nd(
        |req: crate::runtime::HttpRequest| tokio::io::split(req.io_stream).0,
        || input_stream.opaque_object_take_data(),
    ) else {
        anyhow::bail!("Failed to get input_stream");
    };
```
Roughly meas this pseudo code:
```rust
let read_half: typeof (tokio::io::split(HttpRequest(?).io_stream).0)  = input_stream.opaque_object_take_data().unwrap();
```

Renamed it to `valueof_f2_as_typeof_f1` and added comments.

File: src/host_functions/http_listen.rs (hunk 12 above)
Comment (kvinwang) [reply src/host_functions/http_listen.rs]: Good. Done.

File: src/host_functions/http_request.rs (hunk 8 above)
Comment (kvinwang) [reply src/host_functions/http_request.rs]: Good.

File: src/host_functions/timer.rs (hunk 9 above)
Comment (kvinwang) [reply src/host_functions/timer.rs:4]: Added

File: src/service_keeper.rs (hunk 1 above)
Comment (kvinwang) [reply src/service_keeper.rs]: Done

File: tests/bind.ts (hunk 6 above)
Comment (kvinwang) [reply tests/bind.ts]: Done

File: tests/bind.ts (hunk 5 above)
Comment (kvinwang) [reply tests/bind.ts]: Done.

File: tests/devphase.config.ts
Comment (kvinwang) [reply tests/devphase.config.ts:1]: Bumpped the version to 0.6.7.
Looks like no other changes are needed.

File: src/service/resource.rs (hunk 4 above)
Comment (kvinwang) [reply src/service/resource.rs]: Added a `ValueError::RuntimeDropped` for this kind of error.

File: src/lib.rs (hunk 7 above)
Comment (kvinwang) [reply src/lib.rs]: Turned to using anyhow.

File: src/service_keeper.rs (hunk 10 above)
Comment (kvinwang) [reply src/service_keeper.rs:94]: Added

File: examples/incomingFetch.js
Comment (h4x3rotab) [reply examples/incomingFetch.js:1]: In my understanding, `fetch()` is a method to send http request. So `incomingFetch` doesn't sound reasonable to me. Maybe consider `incomingHttpRequest` or `lowLevelHttpServer`?

File: examples/httpListen.js (hunk 3 above)
Comment (h4x3rotab) [reply examples/httpListen.js:1]: OK. I thought by default, chrome and node.js are both providing `inspect` as the log implementation, instead of the original one. Do you think performance is a concern for not using inspect as the default log impl?

File: examples/incomingFetch.js
Comment (kvinwang) [reply examples/incomingFetch.js:1]: Maybe `listeningFetchEvents`? Because that's CF workers gave it the name.
There is already a low level http listening example.

File: examples/httpListen.js (hunk 3 above)
Comment (kvinwang) [reply examples/httpListen.js:1]: > Do you think performance is a concern for not using inspect as the default log impl?

Yes. It's one of the consideration. Another is to be compatible with the original QuickJS behavior.

File: examples/incomingFetch.js
Comment (h4x3rotab) [reply examples/incomingFetch.js:1]: Got it. I didn't research a lot, but I also searched on CF worker's example page and didn't find name like this. Good to use `listeningFetchEvents`

**Review by h4x3rotab — APPROVED** [review]
LGTM
